### ISSUE TITLE
Implement Run Length Encoding (RLE) / Rule End Encoding (REE) support (Epic)

### ISSUE THREAD

**Issue by alamb** [apache/arrow-rs#3520]
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
Arrow has added REE support https://github.com/apache/arrow/pull/14176, similar to dictionary arrays that allow repeated values to be encoded in a space efficient manner that also allows fast processing.

**Describe the solution you'd like**
Implement REE in arrow-rs. Some likely candidate:

- [ ] Support in DataType
- [ ] Support in ArrayData
- [ ] New REE array
- [ ] Support REE in IPC
- [ ] Support REE in cast kernels
- [ ] Support REE in compute kernels

**Describe alternatives you've considered**

**Additional context**

**Comment by tustvold** [comment]
One thing to perhaps give thought to is what kernels we need to facilitate this use-case. For example, a core pattern for dictionaries is to evaluate against the child values array and then `take` the results based on the keys. We possibly need something similar for the REE use-case, i.e. some sort of `take_runs` kernel.

**Comment by alippai** [comment]
Would be adding parquet support in the scope? I’m not sure if parquet has RLE directly, but highly compressed (zstd,snappy,zlib) values can be converted to REE arrays instead of inflating.

**Comment by tustvold** [comment]
Dictionaries are RLE encoded in parquet, so we could theoretically preserve RLE encoding for dictionary arrays, on top of the existing logic to preserve the dictionary encoding. However, the cost and complexity of translating between the two representations is likely to negate much of the potential gains.

> zlib

Do you have a link for this, I'm not sure how you would reliably use the codecs internal RLE coding as it isn't guaranteed to be at a meaningful granularity for the encoded data?

**Comment by alippai** [comment]
@tustvold I agree that the reading performance might not improve but the downstream operations like filter/join/group by (if optimized for REE) would definitely make it worth. Absolutely not a day 0 feature.

The zlib/zstd idea is far fetched, but:
1.) zlib sometimes decides to store data as Z_RLE, might worth checking if this happens with Parquet too: https://optipng.sourceforge.net/pngtech/z_rle.html
2.) I was wondering if LZ77 -> REE is possible without fully decompressing the data, something like posted here (for a different algorithm): https://www.researchgate.net/publication/261072335_From_Run_Length_Encoding_to_LZ78_and_Back_Again

Both of these are super theoretical and needs access to the compression building blocks, the low level api (as you don’t want to ask the lib to decompress the data, but you want to go compressed -> REE to save big)

**Comment by tustvold** [comment]
I'm not sure you can implement ArrayAccessor for REEArray as it doesn't know its value type? This is fine imo, we don't implement ArrayAccessor for DictionaryArray for much the same reason.

Providing an iterator abstraction, similar to TypedDictionaryArray, that downcasts the values and uses ArrayAccessor to "decode" the runs makes sense to me

Most kernels I imagine will need custom logic to handle RunEncodedArrays, e.g. take will need to parse the runs array and compute a new set of runs along with the take indices to apply to the values array. Filter will need to do something similar. 

One thing to be extremely careful of, is to avoid generic code typed on both the key type and the value type - this explodes codegen and has caused a lot of pain with dictionaries

**Comment by tustvold** [comment]
> Would that be a problem?

Yes, wherever possible we should handle the indices and values separately, similarly to how we currently handle dictionaries. Otherwise you catastrophically explode code gen and build times, see https://github.com/apache/arrow-rs/issues/2596 and https://github.com/apache/arrow-datafusion/pull/4999 for some more context. Non-scalar binary kernels are the only ones where I foresee this not being possible, we should gate those behind feature flags.

TypedDictionaryArray is useful for statically typed codepaths, but using it in kernels that must be generated for every combination of types ends up being problematic

**Comment by fabianmurariu** [comment]
Hi, I'm working on [raphtory](https://www.raphtory.com/#) trying to get a query engine off the ground with datafusion. One of the key ingredients would be REE array support, because hopping around graphs can be expressed efficiently with REE. 

Can I start looking into adding support for filtering or is there ongoing work that's required first?

**Comment by alamb** [comment]
> Hi, I'm working on [raphtory](https://www.raphtory.com/#) trying to get a query engine off the ground with datafusion. One of the key ingredients would be REE array support, because hopping around graphs can be expressed efficiently with REE.


That sounds like an awesome project 🙏 

> Can I start looking into adding support for filtering or is there ongoing work that's required first?

That would be great. 

Some key features for arrays are (I am not sure offhand what REE has already, you would have to to some research)
1. Cast support (to convert back/forth to other types
2. Filter support (as you mentioned)
3. IPC support (for arrow flight, for example)

Something else that might be worth looking at is the ability to read REE arrays directly from parquet (and avoid decoding) if you are reading for arrow

You might take a look at the list of items we made for `StringView` to get some inspiration: https://github.com/apache/arrow-rs/issues/5374

**Comment by brancz** [comment]
Looks like IPC and filtering are done, so we can tick those off (haven't checked/used casting yet with REE so can't say yet).

Another thing we just ran into is parquet conversion:

```
Converting RunEndEncodedType to parquet not supported
```

**Comment by ilsley** [comment]
I am considering using Arrow for a project and REE looks pretty useful. It would be nice though to (eventually) have an efficient/performant conversion to and from parquet. Is there a good chance this will come in due course and/or is there a reasonable workaround in the meantime? The REE I envision is for primitive numeric values, without nesting.

**Comment by tustvold** [comment]
> have an efficient/performant conversion to and from parquet

Parquet's RLE is only supported for dictionary indices and repetition levels, as such a direct translation of a primitive RunArray is not possible - see [here](https://github.com/apache/parquet-format/blob/master/Encodings.md#run-length-encoding--bit-packing-hybrid-rle--3). Theoretically one could preserve a `RunArray<DictionaryArray<PrimitiveArray>>` but as alluded to in https://github.com/apache/arrow-rs/issues/3520#issuecomment-1385127898 it would be extremely difficult. 

That being said parquet does have DELTA_BINARY_PACKED which should compress repeated primitives well - see [here](https://github.com/apache/parquet-format/blob/master/Encodings.md#delta-encoding-delta_binary_packed--5). 

I don't know your precise use-case, but I'd strongly encourage steering clear of RunArray unless you really need it - the performance will often be a lot worse, and the support is very limited.

**Comment by ilsley** [comment]
Thanks, that's good to know. I can have very long stretches of nulls or a repeated value (ranging from a few to more than millions of rows). This does indeed compress well in Parquet, but I'm not sure how to best use Arrow for this because the data coming in (not from parquet) is in an interval format with a start and an end (so easily convertible to RLE or REE).

**Comment by alexanderbianchi** [comment]
Hello! Datadog is interested in REE support on par with the `StringView` support. Looking at this thread and the work that was done for `StringView` https://github.com/apache/arrow-rs/issues/5374  it seems like Cast support is one of the main missing pieces but that the Datatype is already implemented.

Is anyone else actively working on this?

**Comment by alamb** [comment]
> Is anyone else actively working on this?

Hi @alexanderbianchi  -- I don't know of anyone. 

However, I think polarsignals @asubiotto / @brancz may be considering it (not sure if they are working on things internally)

**Comment by asubiotto** [comment]
We are not working on REE cast support right now. I think @brancz is focusing on aggregations.

**Comment by brancz** [comment]
At the moment we're just looking at implementing support for aggregations, tracked in https://github.com/apache/datafusion/issues/16011.

**Comment by alamb** [comment]
I have started adding additional tasks / updates to this ticket in case other people want to follow along or help

**Comment by alexanderbianchi** [comment]
Perfect timing, Datadog is working on the cast support right now and soon the compute kernels. We can push those PR's to iterate on soon, couple open questions on our end about what the code should look like (primarily using the dictionary encoding work as an example, as it seems the most similar to REE) that we can hash out after we push. Feel free to point us in a particular direction if there is anything else that needs to be added (including in the datafusion side)

**Comment by alamb** [comment]
Awesome @alexanderbianchi ! FYI @brancz I think is working on this one:
- https://github.com/apache/datafusion/issues/16011

**Comment by ghost** [comment]
### **Potential Road Map of REE epic:**

_My Current Order_:

- [ ] **Support REE in cast Kernels**
- [ ] **Support Binary Expressions**
- [x] **Write Sql page of expected REE array workloads**
- [ ] **Support string functions for REE**

### Support REE in cast kernels: 

Summary of PR:
This PR Adds support for casting to and from RunEndEncoded (REE) arrays within the Arrow Rust (arrow-rs) library.Also expands the casting logic to handle logical types, ensuring that REE arrays can participate in conversions involving logical representations as well._

[Pull Request for REE arrow cast](https://github.com/apache/arrow-rs/pull/7713)



### REE support in binary expressions (arithmetic, comparison):

_Goals:_

- Enable support for arithmetic and comparison expressions (==, !=, <, >, etc.) over RunEndEncodedArray (REE) inputs.
- Support REE participation in:
- Equality (eq, neq)
- Partial equality (grouping, hashing, join keys)
- Arithmetic expressions (where applicable)
- Ensure consistency with how Arrow currently handles DictionaryArray and other nested/logical types in binary expressions.



_Issue faced:_

- What defines equality between REE arrays?
- Should two REE arrays be considered equal if:
- Their logical arrays (i.e., expanded views) are equal?
    - Their physical run-end structure is equal?
- what attributes should be taken into account when taking the hash of REE arrays. Should the same arrow with two different types be considered equal, should they result in the same hash?

EX:

 `REE_1 = run_ends(int64): [4,9,22] values:[3,4,1]`

 `REE_2 = run_ends(int16):[4,9,22] values:[3,4,1]`

should` REE_1 == REE_2 `since their logical representations are the same or should this result in inequality since the datatypes don't match. What about for hashing?

 

### REE support in string functions

_Goals:_

- Enable REE arrays with string values (RunEndEncoded<Index, Utf8> or Utf8View) to work seamlessly with Arrow's string functions like:
- length, concat, substr
- starts_with, ends_with, contains
- Regex-based ops (like, match, etc.)
- String functions operate over the logical representation of REE not physical.

_Issues:_

**off the top of my head id assume theres a performance cost to constantly transforming REE into their logical forms to perform string operations. there may be smarter ways to keep track of the data inside of REE arrays without needing to continuously decode them**

### Write sql page of expected use’s of REE array workloads.

write a sql_logictest file with REE encoded data and try to do basic queries on it (that will likely most/all fail due to lack of support). For example, something like this https://github.com/apache/datafusion/blob/main/datafusion/sqllogictest/test_files/dictionary.slt 

_What ever operations fail, work on implementing those first for REE._

**Comment by ghost** [comment]
[Pull Request](https://github.com/apache/arrow-rs/pull/7713) for Run End Encoding cast

**Comment by ghost** [comment]
hey @alamb @brancz ,when you get a free moment could you take a look at the [REE casting PR](https://github.com/apache/arrow-rs/pull/7713) and [SQL logic test PR](https://github.com/apache/datafusion/pull/16715) as well as [Arithmetic operations for REE](https://github.com/apache/arrow-rs/pull/7933)?

**Comment by alamb** [comment]
Thanks @rich-t-kid-datadog  -- I will try to find the time . Looks like @gabotechs has reviewed it too so it should be a quick review

**Comment by brancz** [comment]
I just realized that converting run-end arrow arrays to parquet is also not supported. I opened https://github.com/apache/arrow-rs/issues/8016 to track this. I probably won't get to working on this very soon, and I don't know yet how involved this will be.

**Comment by vegarsti** [comment]
> I just realized that converting run-end arrow arrays to parquet is also not supported. I opened [#8016](https://github.com/apache/arrow-rs/issues/8016) to track this. I probably won't get to working on this very soon, and I don't know yet how involved this will be.

I've started on #8016 in https://github.com/apache/arrow-rs/pull/8069. So far I have a test where I can write a `RunArray` to Parquet but it's written as plain data, but what we'd want which is dictionary encoded data. I haven't figured out yet where to hook in so that I can somehow convert to dictionary data. Is the work on casting relevant? Pointers welcome!

**Comment by alamb** [comment]
> > I just realized that converting run-end arrow arrays to parquet is also not supported. I opened [#8016](https://github.com/apache/arrow-rs/issues/8016) to track this. I probably won't get to working on this very soon, and I don't know yet how involved this will be.
> 
> I've started on [#8016](https://github.com/apache/arrow-rs/issues/8016) in [#8069](https://github.com/apache/arrow-rs/pull/8069). So far I have a test where I can write a `RunArray` to Parquet but it's written as plain data, but what we'd want which is dictionary encoded data. I haven't figured out yet where to hook in so that I can somehow convert to dictionary data. Is the work on casting relevant? Pointers welcome!

Thanks for the help @vegarsti 

Unfortunately, I don't have much time to devote to REE at the moment (I am slammed with helping along avro and variant, among other things). I can act in a high level, but I can't realistically review the code in any level of detail

@brancz / @rich-t-kid-datadog  -- do you have time to hep review REE PRs, move the project along? I think review capacity is the most limited resource at the moment

**Comment by brancz** [comment]
We're no longer using Parquet as our storage format, so that is no longer a priority for us, and my time is so thinly provisioned right now that I unfortunately cannot afford to spend time on anything that isn't a direct priority. I'll try my best to review everything related to REE aggregations (that includes casting support).

**Comment by albertlockett** [comment]
I'm happy to help with this, especially anything related to supporting REE in parquet

**Comment by alamb** [comment]
> I'm happy to help with this, especially anything related to supporting REE in parquet

Thank you @albertlockett 

One big missing piece I know if is the ability to read data from Parquet as REE arrays -- even though often the pages are compressed with the REE/Bit packing hybrid

So the idea would be to specify to the parquet reader that we wanted to read data as an REE array and then implement the appropriate decoders to do that

Before doing this it would likely be very helpful to create some sort of benchmark / example where reading an REE array directly would help a lot (maybe a column with very many repeated string values 🤔 )

**Comment by vegarsti** [comment]
Some recent progress: Casting should land soon, PR here https://github.com/apache/arrow-rs/pull/8589. Once that is merged, it unblocks the #8069 PR to convert to Parquet.

**Comment by vegarsti** [comment]
@alamb Now that the cast PR has been merged, we can check that bullet in this tracking issue 🎉 

I'm going to continue with the Parquet conversion PR #8069 (which closes issue #8016), and would also be happy to take over the compute kernel work in https://github.com/apache/arrow-rs/pull/7933 to get that over the line as well.

**Comment by alexanderbianchi** [comment]
Hi @vegarsti! Thanks for picking up @rich-t-kid-datadog 's work, he's coming back to Datadog from his internship but 0 need to block any work on that. 
Feel free to pickup the compute kernel work. And my apologies I couldn't carry it on myself after his internship was over 🙇.

**Comment by alamb** [comment]
> [@alamb](https://github.com/alamb) Now that the cast PR has been merged, we can check that bullet in this tracking issue 🎉

Done! Thank you

**Comment by Jefffrey** [comment]
Status update:

I think to close this epic we'd need the following PRs/issues at minimum:

- https://github.com/apache/arrow-rs/issues/8016
- https://github.com/apache/arrow-rs/pull/7933
- https://github.com/apache/arrow-rs/issues/9360

The other issues probably don't need to block closing this epic as completed:

- https://github.com/apache/arrow-rs/pull/8019
- https://github.com/apache/arrow-rs/issues/7693
- https://github.com/apache/arrow-rs/issues/7710
- https://github.com/apache/arrow-rs/issues/8707
- https://github.com/apache/arrow-rs/issues/8042
- https://github.com/apache/arrow-rs/issues/9359

**Comment by Rich-T-kid** [comment]
> > I'm happy to help with this, especially anything related to supporting REE in parquet
> 
> Thank you [@albertlockett](https://github.com/albertlockett)
> 
> One big missing piece I know if is the ability to read data from Parquet as REE arrays -- even though often the pages are compressed with the REE/Bit packing hybrid
> 
> So the idea would be to specify to the parquet reader that we wanted to read data as an REE array and then implement the appropriate decoders to do that
> 
> Before doing this it would likely be very helpful to create some sort of benchmark / example where reading an REE array directly would help a lot (maybe a column with very many repeated string values 🤔 )

I'm working on #8016 and realized the read path (parquet → REE) also isn't currently implemented.

Reading through this thread I'm a bit confused on the right approach. Per the [Parquet spec](https://parquet.apache.org/docs/file-format/data-pages/encodings/), RLE is only used for BOOLEAN values and dictionary indices — whereas in Arrow, an REE array can wrap values of any type.

What I'm really trying to confirm is the decode flow. Normally for a column chunk (e.g. strings), you'd decode each value alongside the rep/def levels to handle nulls etc. But since Parquet's RLE for column data only shows up as `RLE_DICTIONARY` on data pages (which requires a dictionary page), I think the REE read path would look more like:

1. Read the dictionary page to get the unique values.
2. Walk the RLE-encoded indices on the data pages, and for each run, append a `(value, run_end)` pair directly into the REE array — rather than materializing every row first.

Is that the right mental model, or is there a different mapping people have in mind?

This feels distinct enough from the write path in #8016 that it probably warrants its own tracking issue, Im happy to file one.

cc @alamb @albertlockett @Jefffrey @vegarsti

**Comment by Jefffrey** [comment]
I think this sounds in line with the discussion over on the C++ version

- https://github.com/apache/arrow/issues/32339
- https://github.com/apache/arrow/pull/46103

**Comment by Rich-T-kid** [comment]
> Status update:
> 
> I think to close this epic we'd need the following PRs/issues at minimum:
> 
> * [Support converting RunEndEncodedType to parquet #8016](https://github.com/apache/arrow-rs/issues/8016)
> * [Implements Sum,sum_checked,min,max,is Distict,inverse for REE.  #7933](https://github.com/apache/arrow-rs/pull/7933)
> * [Support `RunEndEncoded` in ord comparator #9360](https://github.com/apache/arrow-rs/issues/9360)
> 
> The other issues probably don't need to block closing this epic as completed:
> 
> * [Implement String functions for REE #8019](https://github.com/apache/arrow-rs/pull/8019)
> * [Improve performance of RunArray --> Row conversion #7693](https://github.com/apache/arrow-rs/issues/7693)
> * [Potential Optimization for interleave/take on RunEndEncoded arrays #7710](https://github.com/apache/arrow-rs/issues/7710)
> * [Improve performance of RunEndEncoded cast #8707](https://github.com/apache/arrow-rs/issues/8707)
> * [PrimitiveRunBuilder does not support augmenting its data_type. #8042](https://github.com/apache/arrow-rs/issues/8042)
> * [Support `RunEndEncoded` arrays in `arrow-json` #9359](https://github.com/apache/arrow-rs/issues/9359)

with #8016 being closed, is the Implement RunArray epic considered done?
@alamb @Jefffrey

**Comment by Jefffrey** [comment]
Thanks for getting it over the line @Rich-T-kid 🎉

**Comment by vegarsti** [comment]
Amazing! Thank you and well done @Rich-T-kid!

**Comment by brancz** [comment]
I don’t see “REE in compute kernels” ticked off and I don’t remember it having been implemented either. Should we create a separate issue for it or reopen this one?

**Comment by Jefffrey** [comment]
I believe we have support for most of the kernels last I did a spot check; if we're still missing some we can open issues for them
